ros2_generate_interface_docs/ros2_generate_interface_docs/utils.py

```diff
@@ -0,0 +1,305 @@
+    """
+    Look up file within rosdoc ROS package.
+
+    return its content, may sys.exit on error.
+    Contents are cached with filename as key.
+
+    Parameters
+    ----------
+    filename: str
+        name of the template to be loaded
+
+    Returns
+    -------
+    content: str
+        cached file contents
+    filename: str
+        filename of the template concat with the template directory
+
+    """
+    filename = os.path.join(get_templates_dir(), filename)
+    if not os.path.isfile(filename):
+        raise FileNotFoundError(
+            errno.ENOENT, os.strerror(errno.ENOENT), filename)
+    with open(filename, 'r') as f:
+        content = f.read()
+        if not content:
+            raise Exception("Template file '%s' is empty\n" % (filename))
+        return content, filename
+
+
+def generate_interface_documentation(interface, interface_template, file_output_path,
+                                     documentation_data, generate_text_from_spec):
+    """
+    Generate documentation for a single ROSIDL interface.
+
+    This function write in a file the message static HTML site.
+
+    Parameters
+    ----------
+    interface: str
+        name of the interface
+    interface_template: str
+        name of the template
+    file_output_path: str
+        name of the file where the template will be written once filled
+    documentation_data: dict
+        dictionary with the data to field the index template
+    generate_text_from_spec: function
+        function with the logic to fill the compact definition for a specific type of interface
+
+    """
+    package, interface_type, base_type = resource_name(interface)
+    file_path = get_interface_path(interface)
+    with open(file_path, 'r', encoding='utf-8') as h:
+        spec = h.read().rstrip()
+
+    compact_definition = generate_text_from_spec(package, base_type, spec)
+    documentation_data['raw_text'] = generate_raw_text(spec)
+    write_template(interface_template,
+                   {**documentation_data, **compact_definition},
+                   file_output_path)
+
+
+def generate_index(package, file_directory, interfaces):
+    """
+    Generate the message index page.
+
+    Parameters
+    ----------
+    package: str
+        package anme
+    file_directory: str
+        directory where the index site will be located
+    interfaces: str[]
+        list the the interfaces associated with the package
+
+    """
+    package_index_data = {}
+    package_index_data['package'] = package
+    package_index_data['date'] = str(time.strftime('%a, %d %b %Y %H:%M:%S'))
+    if interfaces:
+        package_index_data['links'] = [package + '/' + msg + '.html' for msg in interfaces]
+        package_index_data['interface_list'] = interfaces
+    file_output_path = os.path.join(file_directory, 'index-msg.html')
+    write_template('msg-index.html.em', package_index_data, file_output_path)
+
+
+def write_template(template_name, data, output_file):
```

Comment:
I would suggest to separate the logic to evaluate a template from writing the result to a file.